### ISSUE TITLE
Linking not working

### ISSUE THREAD

**Issue by Chaphasilor** [rudihorn/drm-vc4-grabber#2]
Hi, I'm coming back to this after some time and wanted to try things again with a "clean slate".  
So I downloaded the latest OSMC beta image and did a dist-upgrade, along with installing Rust via rustup.

However, as soon as I get to the build command, I receive the following errors:

```sh-session
$ cargo build --release --target armv7-unknown-linux-gnueabihf
   Compiling libc v0.2.98
   Compiling crossbeam-utils v0.8.5
   Compiling bitflags v1.2.1
   Compiling crossbeam-epoch v0.9.5
error: linker `/usr/bin/arm-linux-gnueabihf-gcc` not found
  |
  = note: No such file or directory (os error 2)

error: aborting due to previous error

error: could not compile `crossbeam-epoch`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: linker `/usr/bin/arm-linux-gnueabihf-gcc` not found
  |
  = note: No such file or directory (os error 2)

error: aborting due to previous error

error: linker `/usr/bin/arm-linux-gnueabihf-gcc` not found
  |
  = note: No such file or directory (os error 2)

error: aborting due to previous error

error: linker `/usr/bin/arm-linux-gnueabihf-gcc` not found
  |
  = note: No such file or directory (os error 2)

error: aborting due to previous error

error: build failed
```

I didn't get this error last time and I tried to install the linker manually via apt without any success. Did something change, maybe with the OSMC distro? Could you try if it works on your machine using the latest image and upgrades?

I'd love to offer you some more debug info for the original issue, but as of now I'm a bit stuck 😕

**Comment by Daenara** [comment]
I am trying to get this to work, too. It took me a while to even get to the point to be able to do a cargo build command, but then it stopped with the linker, I checked, it was not there. 

After a bit of google I installed aptitude and tried installing that way:
```
osmc@Pi-Kun:~/drm-vc4-grabber$ sudo aptitude install gcc-arm-linux-gnueabi
Note: selecting "gcc-arm-linux-gnueabi:arm64" instead of the virtual package "gcc-arm-linux-gnueabi"
The following NEW packages will be installed:
  binutils-arm-linux-gnueabi:arm64{a} binutils-common:arm64{a} cpp-8-arm-linux-gnueabi:arm64{a} cpp-arm-linux-gnueabi:arm64{a} gcc-8-arm-linux-gnueabi:arm64{ab} gcc-8-arm-linux-gnueabi-base:arm64{a} gcc-8-base:arm64{a} gcc-arm-linux-gnueabi:arm64 libc6:arm64{a} libcc1-0:arm64{a}
  libgcc1:arm64{a} libgmp10:arm64{a} libisl19:arm64{a} libmpc3:arm64{a} libmpfr6:arm64{a} libstdc++6:arm64{a} zlib1g:arm64{a}
The following packages are RECOMMENDED but will NOT be installed:
  libc6-dev-armel-cross libidn2-0:arm64
0 packages upgraded, 17 newly installed, 0 to remove and 0 not upgraded.
Need to get 21.9 MB of archives. After unpacking 85.8 MB will be used.
The following packages have unmet dependencies:
 gcc-8-arm-linux-gnueabi:arm64 : Depends: libgcc-8-dev-armel-cross:arm64 (>= 8.3.0-2cross1) which is a virtual package and is not provided by any available package

The following actions will resolve these dependencies:

     Keep the following packages at their current version:
1)     gcc-8-arm-linux-gnueabi:arm64 [Not Installed]
2)     gcc-arm-linux-gnueabi:arm64 [Not Installed]
```

Apparently libgcc-8-dev-armel-cross:arm64 is the problem. 

I was stumped for a while with all the amd64 stuff, so I checked with uname -a and apparently this is the first pi os in this house to run a 64bit system. I don't know if older OSMC releases were 32 bit, but all my other pi installations were, so that could be the problem. Another thing on the top of my head is the new OSMC release being on debian buster. I do know that the beta releases ran on buster-devel, so maybe there are some packages in there missing now.

Also, I don't have any rust knowledge at all, but aren't linker and toolchain normally for cross-compiling? If so, aren't we trying to compile stuff for the pi on the pi using a cross-compiler for compiling stuff for the pi not on a pi? Shouldn't there be a way to just naturally compile on the pi?

**Comment by Chaphasilor** [comment]
Thanks for the writeup!

About the 64 bit stuff: it's a bit weird with OSMC, iirc they are running a 64 bit kernel with a 32 bit userland.  
It did work though on a previous build, so maybe the problem is a change in the dependency repositories?  
I'm still running the beta build, not the stable one, so should still be on buster-devel.

The linker is always needed, even for native "compilation". We could probably use a different linker somehow, but I don't know how to tell that to cargo 😅  
If @rudihorn could enlighten us about this that would be amazing :D

Also, I still have the SD card with the working linker lying around, but I'm not sure if upgrading would break things there as well.  
We should probably try to find a solution that works with the current build anyway, so I don't think it's super-useful to fiddle around with the working build...

**Comment by Chaphasilor** [comment]
I made sure to set the env var, but haven't tried without specifying the release yet (trying it right now).

Trying to compile from a regular machine sounds like a good idea, you're right. Will try that later as well!

**Comment by Daenara** [comment]
I tried without the target, but the problem is the linker just can't be installed on osmc because one dependency doesn't seem to exist. Aptitude at least tells more than apt-get, apparently it is a virtual package (I saw this once before, but it could resolve that into an actual package) and therefore can't be installed.

Here are all things in /usr/bin that start with a that I actually have, maybe one of those arm things could be used:
![image](https://user-images.githubusercontent.com/3619315/129297251-116e4cef-662c-4bd6-927e-f70bb1e4209b.png)


I normally compile stuff directly on the device I will be using it, and while my pi 1 had me compiling stuff overnight, I haven't found anything that takes me longer than an hour on my pi 4.

**Comment by Chaphasilor** [comment]
@Daenara try `apt-get install gcc-arm-linux-gnueabihf`. I had the same problem with the missing linker in WSL and that fixed it for me. Solution taken from [here](https://stackoverflow.com/questions/28248539/install-cross-compiler-in-debian).

If this also works for you (and maybe even on the Pi), @rudihorn should update the readme with the working command :)

**Comment by Daenara** [comment]
Sadly, this does not work because the package with the virtual package as dependency is a dependency to that one. I might just compile in one of my ubuntu vms to test this, I only need static light in hyperion normally, but my father had to downgrade after the grabbers didn't work anymore so I need to figure out if this runs instead for him, he is a complete linux noob and therefore can't compile it at all.

```
osmc@Pi-Kun:~$ sudo apt-get install gcc-arm-linux-gnueabihf
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 gcc-arm-linux-gnueabihf:arm64 : Depends: gcc-8-arm-linux-gnueabihf:arm64 (>= 8.3.0-1~) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

**Comment by Daenara** [comment]
After asking someone with more linux knowledge for help I got it to compile on my pi4. 

I installed gcc:armhf which included the needed linker. Then I ran into another error with missing .o files but installing libc6-dev fixed that. Now I get the same error that the other issue is discussing (or at least I think so. I added a screenshot of it, if it is another error instead, just tell me and I will open another issue for it.

![image](https://user-images.githubusercontent.com/3619315/129464042-59c06ed3-9d8f-46d8-a069-655d9415207b.png)
